shared/src/balancer_event_handler.rs

```diff
@@ -167,3 +384,256 @@ fn convert_pool_registered(
+        assert_eq!(pool_store.pools_by_token, expected_token_map);
+
+        // Branch where pool registered already stored
+        let pool_address = H160::from_low_u64_be(1);
+        let specialization = PoolSpecialization::General;
+        let mut weighted_pool = WeightedPool {
+            pool_id,
+            pool_address: Some(pool_address),
+            specialization: Some(specialization),
+            tokens: None,
+            block_created: 1,
+        };
+        let mut pools = hashmap! {H256::from_low_u64_be(1) => weighted_pool.clone() };
+        let mut pools_by_token = HashMap::new();
+        let mut pool_store = BalancerPools {
+            pools_by_token: pools_by_token.clone(),
+            pools: pools.clone(),
+            contract_deployment_block: 0,
+        };
+
+        let registration = TokensRegistered {
+            pool_id,
+            tokens: tokens.clone(),
+        };
+
+        pool_store.insert_token_data(index, registration);
+
+        // update expected state
+        weighted_pool.tokens = Some(tokens.clone());
+        *pools.get_mut(&pool_id).unwrap() = weighted_pool;
+        pools_by_token.insert(tokens[0], hashset! { pool_id });
+        pools_by_token.insert(tokens[1], hashset! { pool_id });
+
+        assert_eq!(pool_store.pools, pools);
+        assert_eq!(pool_store.pools_by_token, pools_by_token);
+    }
+
+    #[test]
+    fn balancer_delete_pools() {
+        // Construct a bunch of pools
+        let n = 3;
+        let tokens: Vec<H160> = (0..n + 1).map(H160::from_low_u64_be).collect();
+        let mut pools = HashMap::new();
+        let mut pools_by_token: HashMap<H160, HashSet<H256>> = HashMap::new();
+        for i in 0..n {
+            let pool_id = H256::from_low_u64_be(i + 1);
+            let token_a = tokens[i as usize];
+            let token_b = tokens[i as usize + 1];
+            pools.insert(
+                pool_id,
+                WeightedPool {
+                    pool_id,
+                    tokens: Some(vec![token_a, token_b]),
+                    block_created: i + 1,
+                    // Pool Specialization isn't relevant here.
+                    ..Default::default()
+                },
+            );
+            pools_by_token.entry(token_a).or_default().insert(pool_id);
+            pools_by_token.entry(token_b).or_default().insert(pool_id);
+        }
+
+        let mut pool_store = BalancerPools {
+            pools_by_token,
+            pools,
+            contract_deployment_block: 0,
+        };
+        assert_eq!(pool_store.last_event_block(), 3);
+        pool_store.delete_pools(3).unwrap();
+        assert_eq!(pool_store.last_event_block(), 2);
+        pool_store.delete_pools(2).unwrap();
+        assert_eq!(pool_store.last_event_block(), 1);
+        pool_store.delete_pools(1).unwrap();
+        assert_eq!(pool_store.last_event_block(), 0);
+    }
+
+    #[test]
+    fn balancer_insert_events() {
```

Comment:
You could add a unit test with a reorg where a pool was created but later we replace events without the pool creation (asserting that the pool is no longer there)